frappe/desk/doctype/tag/tag.py

```diff
@@ -34,6 +34,7 @@ def check_user_tags(dt):
 @frappe.whitelist()
 def add_tag(tag, dt, dn, color=None):
 	"adds a new tag to a record, and creates the Tag master"
+	frappe.has_permission(dt, throw=True)
```

Comment:
Pass document name too here. Some permission model works on individual documents.

Reply:
Addressed.

For `add_tags`, validation is thrown if any of the docs is missing permission.